### PR TITLE
[BFD-615] Adds an EFT EFS terraform module for provisioning dedicated EFT file systems

### PR DESCRIPTION
**Pull request by keithdadkins**
### Change Details

This PR adds an EFT EFS module so that we can easily deploy all the resources necessary for BFD to securely share EFT files with our partners using Amazon's Elastic File System (EFS) service. 

The eft_efs module, when called, will:
1. Provision a dedicated EFS file system for the specified partner and environment.
2. Encrypts the file system using a dedicated KMS CMK (customer managed key) that enables encryption-at-rest.
3. Provision an EFS Access Point that our partners will use to mount the file system.
4. Configures the EFS file system policy that:
   1. enforces TLS for all connections (encryption-in-transit)
   2. allows our ETL/pipeline instance to mount the full file system with read/write permissions
   3. allows the partner's account to mount the file system via an access point
   4. blocks all connections not mounted via mount targets (preventing direct NFS access)
5. Creates an IAM role and policy that allows the role to query file systems, mount targets, and access points. Our partners will assume this role in order to query and mount their file systems.
6. Creates security group rules (firewall rules) that allows NFS 2049 (EFS is an enchanced NFS service).
7. Creates a resuable efs_alarms module for cloudwatch monitoring and alerting.

## Notes

 I realize this is a fairly large PR, but most of the files are simply glue for passing variables around terraform. The real *meat* of the PR is in the `ops/terraform/modules/resources/eft_efs/main.tf` file. Things like cloudwatch alarms and settings can be tweaked later on in separate PR's as needed.

**Provisioning EFS**

This PR **does not provision any resources**, but I included a redacted `terraform plan` below so you can see the results. 

If you prefer to run a plan yourself, you will need to configure a partner in the `ops/terraform/modules/stateful/main.tf` file, ensure all the required variables are configured for that partner in the `terraform.tfvars` file, and then run the `terraform plan` from the `ops/terraform/env/<env name>/stateful` directory.

Here is an example configuration for partner "foo" (you could test with 'bcda' instead of 'foo' as they are configured in tfvars, just do not apply):

```terraform
# ops/terraform/modules/stateful/main.tf
...
## EFT EFS
# foo
module "foos_efs" {
  source           = "../resources/eft_efs"
  partner          = "foo"
  partner_acct_num = var.partner_acct_nums["foo"]
  partner_subnets  = var.partner_subnets["foo"]
  posix_gid        = 1500
  posix_uid        = 1500

  env_config = local.env_config
  role       = "etl"
  layer      = "data"
}
```

### Acceptance Validation
- [x] `terraform plan` included in PR for quick review.
- [x] No glaring security concerns.
- [x] Does not contain any potentially sensitive information such as account numbers, subnets, etc. 
- [x] EFT EFS module deployed to lower environments and tested across accounts with a partner.
- [x] PR does not deploy any AWS resources and does not contain any partner specific details.

### Feedback Requested
Any glaring issues, security concerns, etc.

### External References

- [BFD-615](https://jira.cms.gov/browse/BFD-615)
- [BFD-481 (draft SIA approval)](https://jira.cms.gov/browse/BFD-481)
- [Mount helper script (wip) for our partners](https://github.com/CMSgov/beneficiary-fhir-data/pull/586/files)

### Security Implications
The EFT EFS file systems will eventually handle PII/PPHI, introduces new software dependencies (amazon-efs-utils) for the ETL pipeline, and contains altered security controls as outlined in the [(currently in draft) SIA document.](https://confluence.cms.gov/display/BFD/2021-02-03++BFD+EFT+Service+-+Draft?flashId=1007769504). These issues will need to be reviewed and approved before actually deploying and sharing EFT files with our partners (in other PR's). For this work, we just want to ensure that sharing files across accounts with our partners is done as **securely** as possible.

Security Checklist
- [x] EFS file systems are encrypted-at-rest using dedicated CMK's managed by BFD.
- [x] All non-TLS encrypted connections are blocked.
- [x] Access to the EFT EFS file system require IAM authentication and authorization.
- [x] Network access to the EFS file systems restricted to BFD's ETL and partners data subnets only.
- [x] Security group (firewall) rules are in place to restrict network access to EFS/NFS ports only.

### terraform plan
Here is a largely redacted `terraform plan` using "bcda" as a test (in the test environment).

```terraform
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.stateful.module.bcda.aws_efs_access_point.eft will be created
  + resource "aws_efs_access_point" "eft" {
      ...
      + posix_user {
          + gid = <POSIXID>
          + uid = <POSIXID>
        }
      + root_directory {
          + path = "/dropbox"
           ...
              + permissions = "0750"
      ...
    }

  # module.stateful.module.bcda.aws_efs_file_system.eft will be created
  + resource "aws_efs_file_system" "eft" {
      ..
    }

  # module.stateful.module.bcda.aws_efs_file_system_policy.eft will be created
  + resource "aws_efs_file_system_policy" "eft" {
      ...
    }

  # module.stateful.module.bcda.aws_efs_mount_target.eft["subnet-foo1"] will be created
  + resource "aws_efs_mount_target" "eft" {
      ...
    }

  # module.stateful.module.bcda.aws_efs_mount_target.eft["subnet-foo2"] will be created
  + resource "aws_efs_mount_target" "eft" {
      ..
    }

  # module.stateful.module.bcda.aws_efs_mount_target.eft["subnet-foo3"] will be created
  + resource "aws_efs_mount_target" "eft" {
      ..
    }

  # module.stateful.module.bcda.aws_iam_policy.eft_efs_query will be created
  + resource "aws_iam_policy" "eft_efs_query" {
      + description = "Assumed role policy to allow bcda to query BFD file sytems and mount targets"
       ...
    }

  # module.stateful.module.bcda.aws_iam_policy_attachment.eft_efs_query will be created
  + resource "aws_iam_policy_attachment" "eft_efs_query" {
       ...
    }

  # module.stateful.module.bcda.aws_iam_role.eft_efs_query will be created
  + resource "aws_iam_role" "eft_efs_query" {
       ...
       [32m+ assume_role_policy    = jsonencode(
                      + Action    = "sts:AssumeRole"
                      + Effect    = "Allow"
                      + Principal = {
                          + AWS = "arn:aws:iam::1234567890:root"
                      ...
      + description           = "IAM role to allow bcda to query BFD (EFS) file systems and mount targets"
       ...
    }

  # module.stateful.module.bcda.aws_kms_key.eft_efs will be created
  + resource "aws_kms_key" "eft_efs" {
      + description              = "bcda-eft-efs-test-cmk"
      + enable_key_rotation      = true
      + policy                   = jsonencode({
              + Id        = "bcda-eft-efs-test-cmk-policy"
              ...
                      + Action    = "kms:*"
                      + Effect    = "Allow"
                      + Principal = { AWS = "arn:aws:iam::BFDACCTNUM:root" }
              ...
    }

  # module.stateful.module.bcda.aws_security_group.eft_efs_sg will be created
  + resource "aws_security_group" "eft_efs_sg" {
       ...
      + description            = "allows nfs to bcda and bfd subnets"
       ...
    }

  # module.stateful.module.bcda.aws_security_group_rule.bfd_nfs will be created
  + resource "aws_security_group_rule" "bfd_nfs" {
      + cidr_blocks              = [
          + "10.20.30.1/32",
          + "10.20.30.2/32",
          + "10.20.30.3/32",
        ]
      + description              = "Allow NFS"
      + from_port                = <NFSPORTNUM>
      + to_port                  = <NFSPORTNUM>
      + type                     = "ingress"
    }

  # module.stateful.module.bcda.aws_security_group_rule.partner_nfs will be created
  + resource "aws_security_group_rule" "partner_nfs" {
      + cidr_blocks              = [
          + "172.3.4.5/32",
        ]
      + description              = "Allow NFS"
      + from_port                = <NFSPORTNUM>
      + to_port                  = <NFSPORTNUM>
      + type                     = "ingress"
    }

  # module.stateful.module.bcda.aws_sns_topic.eft_efs will be created
  + resource "aws_sns_topic" "eft_efs" {
      + name   = "bcda-eft-efs-test-cloudwatch-alarms"
    }

  # module.stateful.module.bcda.module.cloudwatch_alarms_efs.aws_cloudwatch_metric_alarm.burst_credit_balance_too_low will be created
  + resource "aws_cloudwatch_metric_alarm" "burst_credit_balance_too_low" {
      ...
    }

  # module.stateful.module.bcda.module.cloudwatch_alarms_efs.aws_cloudwatch_metric_alarm.client_connections will be created
  + resource "aws_cloudwatch_metric_alarm" "client_connections" {
      ...
    }

  # module.stateful.module.bcda.module.cloudwatch_alarms_efs.aws_cloudwatch_metric_alarm.percent_io_limit_too_high will be created
  + resource "aws_cloudwatch_metric_alarm" "percent_io_limit_too_high" {
      ...
    }

Plan: 17 to add, 0 to change, 0 to destroy.
```

### CODE REVIEW

**Review by msnook — APPROVED** [review]
👍 BCDA has tested this configuration with BFD and it appears to be functioning as expected!

File: ops/terraform/modules/resources/eft_efs/variables.tf [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+variable "partner_acct_num" {
+  description = "Partners AWS Account Number"
+  type        = string
+}
+
+variable "partner_subnets" {
+  description = "Partner subnets requiring access to EFT EFS file systems"
+  type        = map(list(string))
```
Comment (njdister) [inline ops/terraform/modules/resources/eft_efs/variables.tf:13]: Elsewhere this is defined as `map(map(list(string)))`.  Is this intentional?  Just checking!

File: ops/terraform/modules/resources/eft_efs/variables.tf [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+variable "posix_uid" {
+  description = "High numbered posix id for tracking partner owned files. E.g., 1500, 1501, etc"
+}
+
+variable "posix_gid" {
+  description = "High numbered posix id for tracking partner owned files. E.g., 1500, 1501, etc"
+}
```
Comment (njdister) [inline ops/terraform/modules/resources/eft_efs/variables.tf:22]: Do these values matter or are they arbitrary?

File: ops/terraform/modules/resources/eft_efs/variables.tf (hunk 1 above)
Comment (keithdadkins) [reply ops/terraform/modules/resources/eft_efs/variables.tf:13]: We store all the partners info in a `map(map(list(string)))` (ie in a .tfvars file):
Screen Shot 2021-05-21 at 12 20 51 PM 

But when we are provisioning a specific partner, we only pass their info to the module like so:
Screen Shot 2021-05-21 at 12 15 07 PM

File: ops/terraform/modules/resources/eft_efs/variables.tf (hunk 2 above)
Comment (keithdadkins) [reply ops/terraform/modules/resources/eft_efs/variables.tf:22]: It does not matter as long as they do not fall in any reserved id ranges (varies by os). Usually this is 0-999 for service accounts and real users. Systemd also reserves really high 6000+, so I'm suggesting 1500 to be safe. 

It really does not matter in the end as long as the uid/gid is not in use by either BFD or the partner and we agree to these id's before hand.

File: ops/terraform/modules/resources/eft_efs/variables.tf (hunk 1 above)
Comment (njdister) [reply ops/terraform/modules/resources/eft_efs/variables.tf:13]: Thanks, the module instantiation and tfvars context helps.

**Review by njdister — APPROVED** [review]

**Pull request merged**